### ISSUE TITLE
TODO: CUDA Kernel for Transformer Model

### ISSUE THREAD

**Issue by dyanos** [EleutherAI/oslo#131]

**Comment by dyanos** [comment]
The outcome is from executing torch cross entropy after the ls_cross_entropy_layer.
```
Evaluated in 0.2520465850830078 seconds at ls_cross_entropy
Evaluated in 0.18391203880310059 seconds at torch.nn.functional.cross_entropy
```

This outcome is from the reverse approach.

```
Evaluated in 0.20711636543273926 seconds at torch.nn.functional.cross_entropy
Evaluated in 0.23244929313659668 seconds at ls_cross_entropy
```

The performance measure was obtained by randomly initializing a 100x100 matrix and repeating the process 1000 times.

**Comment by dyanos** [comment]
After moving the Multihead Attention, we measured the performance. There are still things left to match the configuration, etc., but I thought it would be meaningful to compare the speed in the current setting, so I tried it.

We are still verifying the operation of single rather than parallel, so we measured the performance based on single.

Trainer code was used in the same way, and we tested the BertSelfAttention module by writing a branch code that uses the code of lightseq2 and implementing the huging face inside.

The learning data consisted of 5 sentences within 10 to 15 tokens simply for speed comparison. The epoch is 10000.

## Measurement
------------------------------------
The speed measurement is...

1. The time it took to execute the trainer code and finish it.
2. It/s value taken from the progress bar

## Result 
------------------------------------------------------------
Using Huggingface Code
1. total 470.70s, user 7.33s 22.05it/s
2. total 461.50s, user 7.41s 21.83it/s

lightseq2의 Multihead Attention사용(fp32)
1. total 399.77s, user 8.54s - fp32 : 25.27it/s
2. total 400.50s, user 8.13s - fp32 : 25.25it/s

lightseq2의 Multihead Attention사용(fp16)
1. total 385.42s, user 8.33s - fp16 : 26.23

## Discussion
---------------------------------------
1. Contrary to my expectation, there is definitely an effect of speed improvement.
2. It is confirmed that this effect will show a better improvement effect for large dataset even in a small amount.
3. In the above results, it was not a small amount, but the loss also converged faster than expected (to 0.0).
- However, this needs to be reviewed by taking the actual output value.

**Comment by dyanos** [comment]
Loss Graph
------------------------------

![loss](https://user-images.githubusercontent.com/2949252/228719445-a567d947-4f73-45cd-89ed-e1186dcda025.png)

**Comment by hyunwoongko** [comment]
print(model.config.items())

**Comment by dyanos** [comment]
@hyunwoongko

```python
import sentencepiece as spm
# load the vocab and model file of sentencepiece 
sp = spm.SentencePieceProcessor(model_file="../my-pytorch-bert/corpus/1.0-unigram-32000-1.0.model")
line = [line.split("\t")[0] for line in open("../my-pytorch-bert/corpus/1.0-unigram-32000-1.0.vocab", "rt")]
vocab_dit = {v: k for k, v in enumerate(line)}
tokenizer = SPTokenizer("../my-pytorch-bert/corpus/1.0-unigram-32000-1.0.vocab", "../my-pytorch-bert/corpus/1.0-unigram-32000-1.0.model")

cls_id = vocab_dit['[CLS]']
sep_id = vocab_dit['[SEP]']
pad_id = vocab_dit['[PAD]']


# Build datasets from the raw text file
raw_datasets = load_dataset("text", data_files="../raw_data.txt", features=Features({"text": Value(dtype="string", id=None)}))

num_proc = multiprocessing.cpu_count()

def group_texts(examples):
    return tokenizer(examples['text'], max_length=1024, truncation=True)

tokenized_datasets = raw_datasets.map(group_texts, batched=True, num_proc=2)
tokenized_datasets = tokenized_datasets.shuffle(seed=42)

# Pretraining the model
#from transformers import BertConfig, BertForPreTraining, BertForMaskedLM
from transformers import BertConfig
from oslo.transformers.models.bert.modeling_bert import BertForPreTraining, BertForMaskedLM

config = BertConfig(
    vocab_size=32000,
    max_position_embeddings=1024,
    pad_token_id = pad_id,
)

#model = BertForPreTraining(config=config)
model = BertForMaskedLM(config=config)

from transformers import Trainer, TrainingArguments, DataCollatorForLanguageModeling

data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=True, mlm_probability=0.15)

training_args = TrainingArguments(
    output_dir=f"output_{lr_rate}",
    overwrite_output_dir=True,
    num_train_epochs=1000000,
    per_device_train_batch_size=64,
    save_steps=1000,
    save_total_limit=2,
    logging_steps=100,
    learning_rate=lr_rate
)

trainer = Trainer(
    model=model,
    args=training_args,
    data_collator=data_collator,
    train_dataset=tokenized_datasets['train'],
)

print(trainer.train())

trainer.save_model('./saved')
```

**Comment by hyunwoongko** [comment]
1. 정성적/정량적으로 봤을때 이 모델이 잘 작동하는지
2. 코드 정리, 폴더 옮기고, 등등등...
3. 모델 확장.

**Comment by dyanos** [comment]
* 다음은 Loss 추적 그래프입니다. 설정과 CUDA오류 잡은 버전으로 실험했습니다.
![image](https://user-images.githubusercontent.com/2949252/233327158-3dc8b232-faac-49db-83d1-d61b07e1a2de.png)

* tokenizer : bert-base-cased
* dataset : wikipedia (huggingface의 datasets에서 제공)

**Comment by hyunwoongko** [comment]
괜찮네요. 코드 정리 되면 머지해도 될 거 같습니다!
